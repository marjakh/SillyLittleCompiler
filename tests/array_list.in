print("ArrayList test");

let array_list_initial_capacity = 6;
let array_list_growth = 2;

function array_list_create() {
  let a = new Array(array_list_initial_capacity + 2);
  a[0] = 0; // length
  a[1] = 6; // capacity
  return a;
}

function array_list_add(a, elem) {
  let len = a[0];
  let capacity = a[1];
  if (len < capacity) {
    // There's space
    a[len + 2] = elem;
    a[0] = len + 1;
    return a;
  }
  let new_capacity = array_list_growth * len;
  let new_a = new Array(new_capacity + 2);
  new_a[0] = len + 1;
  new_a[1] = new_capacity;
  let i = 0;
  while (i < capacity) {
    new_a[i + 2] = a[i + 2];
    i = i + 1;
  }
  new_a[i + 2] = elem;
  return new_a;
}

function array_list_print(a) {
  let len = a[0];
  let capacity = a[1];
  print("ArrayList");
  print("Length");
  print(len);
  print("Capacity");
  print(capacity);
  print("Elements");
  let i = 0;
  while (i < len) {
    print(a[i + 2]);
    i = i + 1;
  }
}

function array_list_foreach(a, f) {
  let len = a[0];
  let i = 0;
  while (i < len) {
    f(a[i + 2]);
    i = i + 1;
  }
}

function array_list_map(a, f) {
  let len = a[0];
  let i = 0;
  while (i < len) {
    a[i + 2] = f(a[i + 2]);
    i = i + 1;
  }
}

let a = array_list_create();
let i = 0;
while (i < 20) {
  a = array_list_add(a, i * 5);
  i = i + 1;
}
array_list_print(a);

function add_one(elem) {
  return elem + 1;
}

function my_print(elem) {
  print(elem);
}

array_list_map(a, add_one);
array_list_foreach(a, my_print);
