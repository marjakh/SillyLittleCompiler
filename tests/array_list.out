ArrayList test
ArrayList
Length
20
Capacity
24
Elements
0
5
10
15
20
25
30
35
40
45
50
55
60
65
70
75
80
85
90
95
1
6
11
16
21
26
31
36
41
46
51
56
61
66
71
76
81
86
91
96
